src/ATokenNaive.sol

```diff
@@ -0,0 +1,357 @@
+  function getPastVotes(address account, uint256 blockNumber) external returns (uint256);
+}
+
+contract ATokenNaive is AToken {
+  using WadRayMath for uint256;
+  using SafeCast for uint256;
+  using GPv2SafeERC20 for IERC20;
+
+  /// @notice The voting options corresponding to those used in the Governor.
+  enum VoteType {
+    Against,
+    For,
+    Abstain
+  }
+
+  /// @notice Data structure to store vote preferences expressed by depositors.
+  struct ProposalVote {
+    uint128 againstVotes;
+    uint128 forVotes;
+    uint128 abstainVotes;
+  }
+
+  /// @notice Must call castVote within this many blocks of the proposal
+  /// deadline, so-as to allow ample time for all depositors to express their
+  /// vote preferences. Corresponds to roughly 4 hours, assuming 12s/block.
+  uint32 public constant CAST_VOTE_WINDOW = 1200;
+
+  /// @notice Map depositor to deposit amount.
+  mapping(address => uint256) public deposits;
+
+  /// @notice Map borrower to total amount borrowed.
+  mapping(address => uint256) public borrowTotal;
+
+  /// @notice Map proposalId to an address to whether they have voted on this proposal.
+  mapping(uint256 => mapping(address => bool)) private _proposalVotersHasVoted;
+
+  /// @notice Map proposalId to vote totals expressed on this proposal.
+  mapping(uint256 => ProposalVote) public proposalVotes;
+
+  /// @notice The governor contract associated with this governance token.
+  IFractionalGovernor public immutable governor;
+
+  /// @dev Constructor.
+  /// @param _pool The address of the Pool contract
+  /// @param _governor The address of the flex-voting-compatable governance contract.
+  constructor(IPool _pool, address _governor) AToken(_pool) {
+    governor = IFractionalGovernor(_governor);
+  }
+
+  // TODO Is there a better way to do this? It cannot be done in the constructor
+  // because the AToken is just used a proxy -- it won't share an address with
+  // the implementation (i.e. this code).
+  function selfDelegate() public {
+    IVotingToken(governor.token()).delegate(address(this));
+  }
+
+  /// @notice Method which returns the deadline (as a block number) by which
+  /// depositors must express their voting preferences to this Pool contract. It
+  /// will always be before the Governor's corresponding proposal deadline.
+  /// @param proposalId The ID of the proposal in question.
+  function internalVotingPeriodEnd(uint256 proposalId)
+    public
+    view
+    returns (uint256 _lastVotingBlock)
+  {
+    _lastVotingBlock = governor.proposalDeadline(proposalId) - CAST_VOTE_WINDOW;
+  }
+
+  /// TODO how to handle onBehalfOf?
+  /// TODO should this revert if the vote has been cast?
+  /// @notice Allow a depositor to express their voting preference for a given
+  /// proposal. Their preference is recorded internally but not moved to the
+  /// Governor until `castVote` is called. We deliberately do NOT revert if the
+  /// internalVotingPeriodEnd has passed.
```

Comment:
I think we figured we'd just save everyone else the gas by not bothering to check each time if the aToken had voted on the proposal yet

Reply:
Got it, that's fair, I'd put that rationale in a comment cause otherwise it does sound like you'd want to revert to keep state internally consistent when a vote ends (i.e. the expressed prefs saved in AToken matching the upstream votes placed is a nice property, bummer to give it up to save gas, but not a big deal + makes sense)

Reply:
Might not be that expensive to just add a `mapping(uint256 => bool) public hasVoted` storage var and write to it in `castVote` then check in here. I'll give it a try and see how much gas it adds

Reply:
Yeah, as written in the comment above, I think it would be preferable to prevent late voting unless the gas price is significant.

Reply:
Implemented in 161fd2d

Without preventing late voting (snapshot at 11822b1):

```
╭─────────────────────────────────────────────────┬─────────────────┬────────┬────────┬────────┬─────────╮
│ src/ATokenNaive.sol:ATokenNaive contract        ┆                 ┆        ┆        ┆        ┆         │
╞═════════════════════════════════════════════════╪═════════════════╪════════╪════════╪════════╪═════════╡
│ Function Name                                   ┆ min             ┆ avg    ┆ median ┆ max    ┆ # calls │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ castVote                                        ┆ 1592            ┆ 54887  ┆ 67815  ┆ 85715  ┆ 35      │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ expressVote                                     ┆ 2443            ┆ 36757  ┆ 47895  ┆ 48521  ┆ 52      │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
```

Preventing late voting (snapshot at 161fd2d):

```
╭─────────────────────────────────────────────────┬─────────────────┬────────┬────────┬────────┬─────────╮
│ src/ATokenNaive.sol:ATokenNaive contract        ┆                 ┆        ┆        ┆        ┆         │
╞═════════════════════════════════════════════════╪═════════════════╪════════╪════════╪════════╪═════════╡
│ Function Name                                   ┆ min             ┆ avg    ┆ median ┆ max    ┆ # calls │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ castVote                                        ┆ 1570            ┆ 72258  ┆ 88473  ┆ 106373 ┆ 40      │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ expressVote                                     ┆ 739             ┆ 37534  ┆ 50097  ┆ 50758  ┆ 59      │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤

```

Adding late-vote prevention increases `castVote` median gas usage by 30% (20658 gas) and `expressVote` median by ~5% (2202 gas).

In the current gas environment it's not a big deal (~$0.65 assuming 26 gwei/gas, but check my math). Still, I think we probably want to make `castVote` as cheap as possible so as not to disincentivize its being called -- especially if gas + ETH prices go up again. I think not disincentivising `castVote` outweighs better UX for people (or apps) that can't be bothered to check whether or not the aToken has cast its weight before they attempt to express. But curious what you guys think.

Snapshots taken with `$ forge test --gas-report`

Reply:
Thanks for checking this. Adding ~20k gas makes sense, since it's one SSTORE to a fresh slot, but that is a lot. OTOH that method is only called once per proposal so I don't think we'll be too sensitive to price there, and increasing `expressVote` by 5% (the one where IMO we should be more price sensitive) isn't too bad (but also not trivial).

Maybe we leave for now but create an issue track this discussion to come back to it later? It's probably worth getting opinions from e.g. the Aave team here too

Reply:
Tracking: https://github.com/ScopeLift/flexible-voting/issues/19